src/cryptonote_core/tx_pool.cpp

```diff
@@ -900,6 +897,50 @@ namespace cryptonote
     }, false, category);
   }
   //------------------------------------------------------------------
+  bool tx_memory_pool::get_incremental_pool_info(time_t start_time, bool include_sensitive, std::vector<tx_details>& added_txs, std::vector<crypto::hash>& removed_txs) const
+  {
+    CRITICAL_REGION_LOCAL(m_transactions_lock);
+    if ((m_added_txs_start_time != (time_t)0) && (m_removed_txs_start_time != (time_t)0))
+    {
+      if ((start_time <= m_added_txs_start_time) || (start_time <= m_removed_txs_start_time))
+      {
+        // If either of the two lists do not go back far enough it's not possible to
+        // deliver incremental pool info
+        return false;
```

Comment:
`wallet2` now implements to update methods based on transaction pool: the "old" one where it actively gets select pool transactions by separate calls and the "new" incremental one. I could return the whole list of pool txs here easily, but to process them higher up in `wallet2`this would make a third update method necessary, one where the complete list of pool txs comes in through the "get blocks" call.

(Incremental is not possible because we lack the removed txs, and as long as not all daemons around support giving back incremental info we can's simply drop the "old" method.)

Reply:
I disagree with the decision not to return the whole list in this case. It introduces hard-to-reason through edge cases that a caller of this endpoint needs to know to handle (fall back to [make the additional calls to the daemon the "old" way of doing things](https://github.com/rbrunner7/monero/blob/5f77e4b1268b0d262521fb4cc5879ea2cc137912/src/wallet/wallet2.cpp#L2724-L2726)). I expect that wallet2 won't be the only caller of this endpoint, and so it shouldn't be designed in a way that introduces idiosyncrasies that wallet2 handles in a sub-optimal way. Ideally we get rid of the "old" way altogether from the client thanks to the existence of this endpoint, not introduce scenarios that are difficult to see that the client needs to fall back to the old method in order to handle safely.

If the daemon isn't able to respond with incremental info, I think it should fall back to responding with the entire pool, and the client should be able to handle that by updating its local list of pool tx's with that response. Any caller of the endpoint should also be able to handle this case smoothly (daemon with 0 tx's in its pool goes offline, comes back online, client makes the request to the daemon).

By returning all tx's in the pool over this endpoint, you can also remove [this excessive call to the daemon on first connection](https://github.com/monero-project/monero/pull/8076#discussion_r809996070), which I think is a pretty significant drawback to this PR as is. Reducing trips to the daemon is one of the main goals of this PR, and the first connection is by far the most important one because it actually impacts UX.

____

It's also generally hard to reason through the existence of 2 tracking variables that are there to essentially solve the same thing: `m_added_txs_start_time` AND `m_removed_txs_start_time`. I think there is room to consolidate what these 2 separate variables are doing into a single variable. If you can manage to simplify the approach along these lines, I think it would make reviewing this code for safety issues a lot easier

Reply:
These are all very reasonable arguments, but I have a question before I start to implement this:

This is a breaking change. A "new" `wallet2` won't be able any more to talk to an "old" daemon. Obviously, the old daemon won't understand the request to give back any pool data together with block data and will give back nothing. If we remove the "old" method of updating the pool it's game over at this point already, and `wallet2` will have to give back an error that it can't use that daemon and disconnect.

I was working under the assumption that this approach does **not** fly. I assumed that introducing such a radical breaking change is unacceptable, and that I **have** to keep `wallet2` able to work with "old" daemons, at least for the time being. I therefore saw keeping two different update systems as unfortunate collateral damage, so to say.

What's your opinion on this? Do you vote for this breaking change?

Of course this question is connected with the hardfork. If it's guaranteed that this PR will make it into the new release, a breaking change is basically nothing to fear: Obviously there can't be any "old" daemons around anymore in this case. But if it misses the hardfork - after all so far nobody assumes this to be an essential part of the hardfork release in this sense yet - it will be in a very difficult situation.

Reply:
I think you were right to implement it with that assumption! I don't think what I'm suggesting necessitates a breaking change, but maybe I'm missing something.

When the client sets `include_pool_info` to true (which it should be able to do on the first request even if talking to an old daemon) and the client is talking to a new daemon that can see that this is true, the daemon should respond with the pool info. I don't think the new daemon should expect that a new client has a fallback to the old logic that correctly handles hard-to-see edge cases.

I think you're right that the client should *also* have fallback logic to handle the circumstance where it is talking to an old daemon (when the daemon response does not include pool info), at least until *after* there is a hard fork with this change included (hopefully this one). Ideally the updated endpoint is implemented in a way that does not require changes to the endpoint's behavior in the future, and we could theoretically remove the old logic from the client *after* we're certain daemons have all updated. Upon re-reading, I was definitely unclear on this

Reply:
Thanks for the clarification. That's reassuring.

Fact remains that if the daemon gives back the whole pool it's **not** an increment, and I will go through the code carefully to see how many changes would really be needed to process this "non-increment" because so far things are programmed under the assumption that it's in fact an increment.

Usually it's a good idea to be very careful with such semantics. I compare this with a call that sometimes gives back a point in time and sometimes a duration since some known point back in time, and you for sure can't treat both the same way, if you know what I mean.

But maybe the needed changes are smaller than I fear, maybe you can indeed treat the whole pool more or less like an especially large increment and most things will just turn out right.

I will probably start to work on this during the coming weekend.

Reply:
@j-berman argued:

> It's also generally hard to reason through the existence of 2 tracking variables that are there to essentially solve the same thing: m_added_txs_start_time AND m_removed_txs_start_time.

I was just going to consolidate into 1 variable when I saw a reason to keep those two:

The list of txs deleted from pool never shrinks in a "natural" way, just more and more entries get added over time. To prevent that list from becoming too big, there is a mechanism to restrict its size, check method `tx_memory_pool::track_removed_tx`. As soon as elements are deleted from the list, the list of added txs and the list of deleted txs don't go back anymore to the same point in time.

Of course it would be possible to adjust a single variable accordingly after deleting in a correct way, but I somehow doubt whether that would be easier to understand than having two variables.

Personally I lean towards keeping two, because now each of the two has a crystal-clear semantic, and I only have to consider both in a single place, when deciding whether the lists go back as far as a particular clients needs them, but if you see it differently, I am ready to modify.

What do you think?

Reply:
@j-berman commented:

> I disagree with the decision not to return the whole list in this case.

You were right to disagree and insist on improvement. I recognized that I had looked at the capabilities of that RPC call in the much too narrow context of only this use case, among other things. It makes perfect sense to return the whole pool if for whatever reason returning incremental info is not possible, and teach `wallet2` to update its internal data from that.

This is fully implemented now.

The complexity of the code went up another notch, because as feared updating from the whole pool returned by the `getblocks` RPC call is technically **not** identical to the "old" method, but it's not too bad, and it's worth it because now already the initial fetching of pool info immediately after opening a wallet can profit from the reduction of RPC calls.